exonum/src/helpers/merkle.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Merkle tree utilities.
+
+use crypto::{self, Hash, HashStream};
+
+/// Computes Merkle root hash for a given list of hashes.
+///
+/// If `hashes` are empty then `Hash::zero()` value is returned.
+pub fn root_hash<'a, I: IntoIterator<Item = &'a Hash>>(hashes: I) -> Hash {
+    let mut current_hashes: Vec<Hash> = hashes.into_iter().cloned().collect();
```

Comment:
Cloning here can be avoided, if the function signature will be changed:
```
pub fn root_hash(hashes: &[Hash]) -> Hash {
    if hashes.is_empty() {
        return Hash::zero();
    }
    let mut current_hashes = combine_hash_list(hashes);
    while current_hashes.len() > 1 {
        current_hashes = combine_hash_list(&current_hashes);
    }
    current_hashes[0]
}
```

What's the profit from taking `IntoIterator` as argument, if it's get collected in the first line? It could be collected as well on the callee side as well.
And with `&[Hash]` code looks easier as well.

Reply:
I used `IntoIterator` as a way to generalize root hash computation, to not limit it to just slices. Though, I guess we can revert this change and keep it with slices until we really need it.

Reply:
As for in-place computation, I guess we can rewrite `combine_hash_list ` to be `fn(&mut Vec)` with in-place computations and zero additional Vec allocations. However, in this case we'd still need to clone the original slice. It is possible to (effectively) avoid copying by constructing the initial vector of the pre-combined half of the hashes, but I don't think it's worth the readability hit.